### DISCUSSION TITLE
I'm facing the following CORS issue

### DISCUSSION THREAD

**Discussion by silverfantacy** [fief-dev/fief#293] [Q&A]
I'm facing the following CORS issue in my development environment using Vite, Vue3 and "@fief/fief" version "^0.15.0-beta.1". How can I resolve it?

Access to fetch at 'https://xxxxfief.dev/.well-known/openid-configuration' from origin 'http://localhost:5173' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Comment by silverfantacy** [comment]
when I execute fiefAuth.authCallback(url). The following error occurs:

Error
at _Fief.<anonymous> (client.ts:824:15)
at Generator.throw (<anonymous>)
at rejected

However, there doesn't seem to be a 400 status code in the network connection.

**Comment by frankie567** [comment]
It's strange that you get such a CORS error when accessing the `/.well-known/openid-configuration` endpoint: we explicitly allow any cross-origin request for this endpoint.

Are you sure you point to the right Fief domain? A correct address looks like: `https://example.fief.dev`

**Comment by meteoDaniel** [comment]
I am facing the same issue using Flask as a backend for a plotly-dash app. 

I still added flask-cors to the flask server. 

Where can I set the suggested request mode to 'no-cors'?

![Bildschirmfoto vom 2024-05-18 14-20-54](https://github.com/fief-dev/fief/assets/27021858/f22e5c96-b936-4627-b387-2860491e015e)

![Bildschirmfoto vom 2024-05-18 14-34-35](https://github.com/fief-dev/fief/assets/27021858/fd5fbfcc-496d-4e2c-b094-5d50bbd66146)

**Reply by meteoDaniel** [reply]
@frankie567  As far as I understand it, this cors happens when I try to open the web-app and the user is not verified at this moment. The decorator implemented followed by your Flask guide, redirects from the App to the fief login portal. 

Redirecting atfer login to the web-app works fine when I open the first url mentioned in the cors error message.

**Reply by frankie567** [reply]
It looks to me that you're trying to get the `/authorize` page from a fetch request in JavaScript code. That won't work: this page is not an API you can request programmatically. You should redirect your user to it so they can go through the login flow directly.

**Reply by meteoDaniel** [reply]
@frankie567 thanks for your answer. 

I followed the guide in your documentation. I have adapted the errorhandler for FiefUnauthorizedError : 

```
@server.errorhandler(FiefAuthUnauthorized)
def fief_unauthorized_error(e):
    return flask.redirect(f"https://fief.alitiq.com/solar/authorize?response_type=code&client_id={os.environ['FIEF_CLIENT_ID']}&redirect_uri=http://localhost:443/overview&scope=openid")

```

The call of the error handling comes rom this callback: 

```
@app.callback(Output("page-content", "children"), [Input("url", "pathname")])
@auth.current_user()
def display_page(pathname):
    """main window content handler"""
    if not pathname in list(LAYOUT_MAPPINGS.keys()):
        raise PreventUpdate

    return LAYOUT_MAPPINGS[pathname]
```

Layout Mapping returns different plotly dash html Objects according to the given path in the url. This callback and the way plotly dash is executing them might be the problem?

**Reply by frankie567** [reply]
I'm not really familiar with Dash but yes, as I understand it, it tries to update the UI by making JS requests and "patching" the DOM. Thus, a redirection response won't be properly handled by Dash.

I quickly checked and maybe you can sort this out with [`dcc.Location`](https://dash.plotly.com/dash-core-components/location):

```py
import dash_core_components as dcc
from flask import Flask, g, redirect, request, session, url_for

@app.errorhandler(FiefAuthUnauthorized)  
def fief_unauthorized_error(e):
    redirect_uri = url_for("auth_callback", _external=True)  
    auth_url = fief.auth_url(redirect_uri, scope=["openid"])  # Use the Fief client to generate URL instead of manually building it yourself
    return dcc.Location(auth_url, refresh=True)  # Tell Dash to make a Redirect?
```

**Reply by meteoDaniel** [reply]
PS: When I login using the link in the browser console, everything works fine with the decorator.

**Reply by meteoDaniel** [reply]
@frankie567 , good idea. It does not work so far, but I think the Error will help understand where the problem is: 

```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 2529, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1826, in full_dispatch_request
    return self.finalize_request(rv)
  File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1845, in finalize_request
    response = self.make_response(rv)
  File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 2174, in make_response
    raise TypeError(
TypeError: The view function did not return a valid response. The return type must be a string, dict, list, tuple with headers or status, Response instance, or WSGI callable, but it was a Location.
```
So the error handler is running make_response()

**Reply by frankie567** [reply]
Hmm, maybe Dash doesn't play well with error handlers. A solution I see could be to mark the current user as `optional` and handle the redirect in the callback:

```py
@app.callback(Output("page-content", "children"), [Input("url", "pathname")])
@auth.current_user(optional=True)  # Optional user
def display_page(pathname):
    """main window content handler"""
    # Manually redirect if no user
    if g.user is None:
        redirect_uri = url_for("auth_callback", _external=True)  
        auth_url = fief.auth_url(redirect_uri, scope=["openid"])  # Use the Fief client to generate URL instead of manually building it yourself
        return dcc.Location(auth_url, refresh=True)  # Tell Dash to make a Redirect?

    if not pathname in list(LAYOUT_MAPPINGS.keys()):
        raise PreventUpdate

    return LAYOUT_MAPPINGS[pathname]
```

**Reply by meteoDaniel** [reply]
Still had this idea but did not knew how to suppress the call of the error handler. 

So far I can clarify that all works fine when adding such a helper: 

```
@app.server.route('/login')
@auth.current_user()
def index_redirect():
    return flask.redirect('/overview')
```

In that case the redirection to fief is running. 

Will be back soon to report about your suggestion.

**Reply by meteoDaniel** [reply]
So the redirection runs right now 
```python
@app.callback([Output("page-content", "children"), Output("url", "href")], [Input("url", "pathname")])
@auth.current_user(optional=True)
def display_page(pathname):
    """main window content handler"""
    if not pathname in list(LAYOUT_MAPPINGS.keys()):
        raise PreventUpdate

    if flask.g.user is None:
        return no_update, f"https://fief.alitiq.com/solar/authorize?response_type=code&client_id={os.environ['FIEF_CLIENT_ID']}&redirect_uri=http://localhost:443/auth-callback&scope=openid"

    return LAYOUT_MAPPINGS[pathname], no_update
```
I had to adapt the redirect_uri to the /auth-callback otherwise I was stuck in an infinte redirect loop. 


Thanks a lot for your support ! As far as we have revenue on the app, I will give some support via polars. 

Best regards
